### ISSUE TITLE
Custom loss reconstruction for Tweedie/Regression_l1 loss

### ISSUE THREAD

**Issue by hilade** [microsoft/LightGBM#6160]
## Description

I am working on a project where I need to implement a custom loss function for the LGBMRegressor model.
Specifically, I am interested in reconstructing (as a first step) the Tweedie or Regression_l1 loss that is being used in the model as an objective. This would allow me to have more flexibility in customizing the loss function while retaining the model's compatibility with existing LightGBM functionalities.

I tried looking at the C++ implementation of [LightGBM#RegressionTweedieLoss()](https://github.com/microsoft/LightGBM/blob/fcf76bceb902a69e1af7cdeac289d31916ddc3f5/src/objective/regression_objective.hpp#L714-L749) and implement the function in python.
I expected my custom loss function to produce results equivalent to using Tweedie or Regression L1 loss in the objective hyperparameter but the results were very different (+- 100% difference). 

## Reproducible example

```
LGBMRegressor(boosting_type='gbdt',
                             num_leaves=1000,
                             max_depth=20,
                             learning_rate=0.01,
                             n_estimators=100,
                             objective=customLossTweedie, # or objective=customLossRegression_l1
                             min_split_gain=0.03,
                             min_child_samples=400,
                             subsample=0.5,
                             colsample_bytree=0.8,
                             reg_alpha=0.5,
                             reg_lambda=0.5,
                             nthread=-1,
                             random_state=10)

def customLossTweedie(y_true, y_pred):
    rho = 1.5
    exp_1_score = np.exp((1 - rho) * y_pred)
    exp_2_score = np.exp((2 - rho) * y_pred)
    grad = -y_true * exp_1_score + exp_2_score
    hess = -y_true * (1 - rho) * exp_1_score + (2 - rho) * exp_2_score
    return grad, hess

def customLossRegression_l1(y_true, y_pred):
    diff = y_pred - y_true
    grad = np.sign(diff)
    hess = np.ones_like(y_pred)
    return grad, hess
```

## Additional Information

LightGBM Version: 3.3.5
Programming Language: Python
Operating System: macOS

Any guidance or examples on how to achieve this customization would be greatly appreciated.
Thank you for your assistance!

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

Also...I see that you double-posted this here and on Stack Overflow ([link](https://stackoverflow.com/questions/77365771/custom-loss-function-of-tweedie-regression-l1-objective-in-lgbmregressor-python)).

Please do not do that. Maintainers here also monitor the `[lightgbm]` tag on Stack Overflow. I could have been spending time preparing an answer here while another maintainer was spending time answering your Stack Overflow post, which would have been a waste of maintainers' limited attention that could otherwise have been spent improving this project. Double-posting also makes it less likely that others with a similar question will find the relevant discussion and answer.

Since you're already getting answers on Stack Overflow, I'm going to close this and encourage anyone who can help to please go answer there.
